packages/root/src/Root.tsx

```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { HTMLElementProps } from '@leafygreen-ui/lib';
+
+type ButtonProps<T> = HTMLElementProps<'button'> &
+  T & {
+    href?: undefined;
+  };
+
+type Anchor<T> = HTMLElementProps<'a'> &
+  T & {
+    href: string;
+  };
+
+type CustomElementProps<T> = T & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type RootProps<T> = ButtonProps<T> | Anchor<T> | CustomElementProps<T>;
+
+function usesCustomElement<T>(
+  props: RootProps<T>,
+): props is CustomElementProps<T> {
+  return (props as any).as != null;
+}
+
+function usesLinkElement<T>(props: RootProps<T>): props is Anchor<T> {
+  return props.href != null;
+}
+
+/**
+ * # Root
+ *
+ * Root component
+ *
+ * ```
+<Root href="https://mongodb.design">Anchors Away!</Root>
+```
+ * @param props.children Takes priority in determining what tag the `<Root />` element will render.
+ * @param props.href Determines the location that an anchor tag will point to. If prop is set without the `as` prop, `<Root />` will render `a` tags.
+ * @param props.as Callback to change the open state of the Menu.
+ */
+function Root<T>(props: RootProps<T>) {
+  const { children, ...rest } = props;
```

Comment:
We should strip off any props that are specific to `Root<T>` only (i.e. `as`). Otherwise they'll be passed in when we use `{...rest}`.

```suggestion
  const { children, as, ...rest } = props;
```
and then just update the reference below from `props.as` to `as`.

Reply:
You can't destructure `as` without Typescript complaining, because the prop won't exist unless we're using a custom element so we have to check that it exists first.

Reply:
Taken offline, but we can use Omit to peel off the property right before we actually spread the rest.